### ISSUE TITLE
JS class style?

### ISSUE THREAD

**Issue by nfarina** [KhaosT/HAP-NodeJS#85]
HAP-NodeJS uses a style of `require`ing and `new`ing objects that I haven't seen before. Each module exports a dictionary of constructors, and modules that use constructors to make classes use a "factory" (?) pattern:

```js
var characteristic_Factor = new require("./Characteristic.js"); // new is used here
[later]
var characteristic = new characteristic_Factor.Characteristic(...); // and here
```

Meanwhile, classes seem to "hijack" their initializer in a strange way:

```js
function Characteristic(options, onUpdate) {
	if (!(this instanceof Characteristic))  {
		return new Characteristic(options, onUpdate); // repeated init options
	}
	... rest of normal init ...
```

I'm curious if there is an advantage to this pattern versus "normal" javascript classes?

**Comment by KhaosT** [comment]
@nfarina do you have an example of "normal" javascript class? I'm pretty new to JavaScript and that's the way I found people use in other repos...

**Comment by nfarina** [comment]
Gotcha! Totally understand. I would personally copy from a really popular NodeJS lib like Express, it uses the traditional JS class structure and exposes the constructor function directly: https://github.com/strongloop/express/blob/master/lib/router/route.js

**Comment by nfarina** [comment]
Would you mind if I submitted a PR to for this? I also might take a stab at documenting some of the classes in the headers (and you can tell me when I guessed wrong).

**Comment by maddox** [comment]
Maybe let #86 land first tho? :sweat_smile:

**Comment by maddox** [comment]
Docs would be :heart::heart::heart:

**Comment by KhaosT** [comment]
@nfarina feel free to do so :)

**Comment by snowdd1** [comment]
> Meanwhile, classes seem to "hijack" their initializer in a strange way:
> 
````
function Characteristic(options, onUpdate) {
     if (!(this instanceof Characteristic))  {
         return new Characteristic(options, onUpdate); // repeated init options
     }
     ... rest of normal init ...
`````
> 
> I'm curious if there is an advantage to this pattern versus "normal" javascript classes?

I think I finally understood that pattern - by reading "JavaScript - the good parts" book to eventually learn the basics. In the authors explanations about classical versus prototypal inheritance, he states that constructors in JS are not secured against accidental calls as "normal functions", but then <code>this</code> is bound to the global object and not to an new instance of the class, meaning all <code>this.variable=value</code> assignments refer to new global variables. The pattern above protects against missing <code>new</code> statements, by returning a new instance and calling the constructor (themselves) again.
What do you think?

**Comment by nfarina** [comment]
Well to me it seems like an odd pattern, since it creates an ambiguous way of instantiating classes. Also, ES6 has removed the ability to drop the `new` keyword like this: http://stackoverflow.com/questions/30689817/es6-call-class-constructor-without-new-keyword

But if you guys feel strongly about it then I can certainly live with it :)

**Comment by snowdd1** [comment]
It is odd at best, but I didn't knew of the class keyword mentioned in the article either. Does JS support classical inheritance, too?

**Comment by nfarina** [comment]
So that's for ES6, the next version of JS. It will be supported in Node soon enough (io.js supports it via babel but it's slated to be merged back into NodeJS core).

**Comment by nfarina** [comment]
Closed by #110
